hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/compact/HoodieFlinkCompactor.java

```diff
@@ -218,74 +228,112 @@ private void compact() throws Exception {
+          : conf.getInteger(FlinkOptions.COMPACTION_TASKS);
 
-      LOG.info("Start to compaction for instant " + compactionInstantTime);
+      LOG.info("Start to compaction for instant " + compactionInstantTimes);
 
       // Mark instant as compaction inflight
-      table.getActiveTimeline().transitionCompactionRequestedToInflight(instant);
+      for (HoodieInstant instant : instants) {
+        table.getActiveTimeline().transitionCompactionRequestedToInflight(instant);
+      }
       table.getMetaClient().reloadActiveTimeline();
 
-      env.addSource(new CompactionPlanSourceFunction(compactionPlan, compactionInstantTime))
-          .name("compaction_source")
-          .uid("uid_compaction_source")
-          .rebalance()
+      // use side-output to make operations that is in the same plan to be placed in the same stream
+      // keyby() cannot sure that different operations are in the different stream
+      Pair<String, HoodieCompactionPlan> firstPlan = compactionPlans.get(0);
+      DataStream<CompactionPlanEvent> source = env.addSource(new CompactionPlanSourceFunction(firstPlan.getRight(), firstPlan.getLeft()))
+          .name("compaction_source " + firstPlan.getLeft())
+          .uid("uid_compaction_source " + firstPlan.getLeft());
+      if (compactionPlans.size() > 1) {
+        for (Pair<String, HoodieCompactionPlan> pair : compactionPlans.subList(1, compactionPlans.size())) {
+          source = source.union(env.addSource(new CompactionPlanSourceFunction(pair.getRight(), pair.getLeft()))
+              .name("compaction_source " + pair.getLeft())
+              .uid("uid_compaction_source_" + pair.getLeft()));
+        }
+      }
+
+      SingleOutputStreamOperator<Void> operator = source.rebalance()
           .transform("compact_task",
               TypeInformation.of(CompactionCommitEvent.class),
               new ProcessOperator<>(new CompactFunction(conf)))
           .setParallelism(compactionParallelism)
-          .addSink(new CompactionCommitSink(conf))
-          .name("clean_commits")
-          .uid("uid_clean_commits")
+          .process(new ProcessFunction<CompactionCommitEvent, Void>() {
+            @Override
+            public void processElement(CompactionCommitEvent event, ProcessFunction<CompactionCommitEvent, Void>.Context context, Collector<Void> out) {
+              context.output(new OutputTag<>(event.getInstant(), TypeInformation.of(CompactionCommitEvent.class)), event);
+            }
+          })
+          .name("group_by_compaction_plan")
+          .uid("uid_group_by_compaction_plan")
           .setParallelism(1);
 
-      env.execute("flink_hudi_compaction_" + compactionInstantTime);
+      compactionPlans.forEach(pair ->
+          operator.getSideOutput(new OutputTag<>(pair.getLeft(), TypeInformation.of(CompactionCommitEvent.class)))
+              .addSink(new CompactionCommitSink(conf))
```

Comment:
no change needed for the CompactionCommitSink?

Reply:
Yes, CompactionCommitSink is designed to handle one compaction plan' status. CompactionCommitSink would collect every compaction operation' status until every operation is complete, and then mark the compaction plan to complete. Second, file id that is need to compacted is not allow be scheduled to different plan, means that the compaction of one plan would not interfere with another plan. So that it is only need to make sure the operation that belong to the same plan into the same CompactionCommitSink.
This is why i use side-output to put operations that belong to the same compaction plan into the same sink.

Reply:
We have a list of compaction plan for the source function, then we can pass the plans to the sink function as well. I think this way is more suitable for batch compaction. The current `CompactionCommitSink` need to handle streaming job scenario as well. Do you think we can have an independent CompactionCommitSink for batch job to have a better fault tolerance? WDYT?

Reply:
> We have a list of compaction plan for the source function, then we can pass the plans to the sink function as well. I think this way is more suitable for batch compaction. The current `CompactionCommitSink` need to handle streaming job scenario as well. Do you think we can have an independent CompactionCommitSink for batch job to have a better fault tolerance? WDYT?

Your suggestion looks well. I have tried to make a new `MultiCompactionSource` and `MultiCompactionSink` to handle multi compaction plan in batch, but i found that the new class is nothing different except a `for` loop. I think flink `side-output` have helped me to finish the `for` loop well. Also, if i put all compaction plan in one sink, one of failure plan would effect other plans, but in fact these plans have no relationship. So it is unneccessary to provide new `MultiCompactionSource` and `MultiCompactionSink` for it in my opinion.

Reply:
In my experience, we need to execute the compaction plan always in FIFO sequence, cause if we execute the later plan first, the log file might be clean and the first plan will get stuck. Once happen, we need to involve human to solve the problem. So I think when we execute multi plan at once, we also need to keep the sequence and commit one by one.

Reply:
It's isolated between each compaction plan, and the log would not be remove before finishing the first compaction plan. So i can compact any compaction plan, and this is why we own the option `--seq` to choose `FIFO` or `LIFO` for compaction.